binder/src/main/java/io/grpc/binder/internal/FlowController.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.grpc.binder.internal;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/** Keeps track of the number of bytes on the wire in a single direction. */
+final class FlowController {
+  private final int maxUnackedBytes;
+
+  @GuardedBy("this")
+  private long totalBytesSent;
+
+  @GuardedBy("this")
+  private long totalBytesAckedByPeer;
+
+  // @GuardedBy("this") for writes but not reads.
+  private volatile boolean transmitWindowFull;
+
+  /**
+   * Creates a new instance of {@link FlowController}.
+   *
+   * @param maxUnackedBytes a weak limit on the number of bytes sent but not yet acknowledged
+   */
+  public FlowController(int maxUnackedBytes) {
+    this.maxUnackedBytes = maxUnackedBytes;
+  }
+
+  /**
+   * Returns true iff the number of reported bytes sent but not yet acknowledged is greater than or
+   * equal to {@code maxUnackedBytes}.
+   */
+  public boolean isTransmitWindowFull() {
+    return transmitWindowFull;
+  }
+
+  /**
+   * Informs this flow controller that more data has been sent.
+   *
+   * @param numBytesSent a non-negative number of additional bytes sent
+   * @return true iff this report caused {@link #isTransmitWindowFull()} to transition to true
+   */
+  public synchronized boolean notifyBytesSent(long numBytesSent) {
+    totalBytesSent += numBytesSent;
```

Comment:
Since this is called for every transaction (even tiny ones), I have a mild preference for AtomicLong & only synchronising within the condition, so we retain the same performance characteristics.

It's possible this was a premature optimisation in the past though, so I could be convinced otherwise.

Reply:
Hmm can you say more about how this would work? Without synchronization, I'm not sure how to reason about the correctness of evaluating the condition. In particular, our view of totalBytesAckedByPeer could be stale (the original race condition we're aiming to solve in this PR)

Reply:
Oh sorry, I missed some details there. We'd need to check again within the internal synchronized block and make totalBytesAckedByPeer volatile.

That said though, perhaps it's best to just stick with synchronization, and optimise later if we see a problem.

Reply:
Sounds good. We can revisit the double checked locking idea once we have a profile that shows it's slow and a stress test to give us confidence in correctness.